### ISSUE TITLE
occasional "socket hang up" errors in APM server intake requests

### ISSUE THREAD

**Issue by trentm** [elastic/apm-agent-nodejs#2594]
We observe occasional "socket hang up" errors in the APM agent's intake requests to APM server. With trace logging, it looks like this:

```
{"log.level":"trace","@timestamp":"2022-03-02T23:53:24.059Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","reqId":"215419e0042e25919a4a32f6b74b1f4c","ecs":{"version":"1.6.0"},"message":"intake request start"}
{"log.level":"trace","@timestamp":"2022-03-02T23:53:24.059Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","encodeTimeMs":0.139534,"fullTimeMs":0.711256,"numEvents":1,"numBytes":775,"ecs":{"version":"1.6.0"},"message":"_writeBatch"}
{"log.level":"trace","@timestamp":"2022-03-02T23:53:24.059Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","activeIntakeReq":true,"lambdaEnd":true,"ecs":{"version":"1.6.0"},"message":"_writeFlush"}
{"log.level":"trace","@timestamp":"2022-03-02T23:53:24.060Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","reqId":"215419e0042e25919a4a32f6b74b1f4c","ecs":{"version":"1.6.0"},"message":"intakeReq \"error\""}
{"log.level":"trace","@timestamp":"2022-03-02T23:53:24.060Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","reqId":"215419e0042e25919a4a32f6b74b1f4c","concluded":false,"ecs":{"version":"1.6.0"},"error":{"type":"Error","message":"socket hang up","stack_trace":"Error: socket hang up\n    at connResetException (internal/errors.js:639:14)\n    at Socket.socketCloseListener (_http_client.js:449:25)\n    at Socket.emit (events.js:412:35)\n    at TCP.<anonymous> (net.js:686:12)\n    at TCP.callbackTrampoline (internal/async_hooks.js:130:17)"},"message":"completePart intakeReq"}
{"log.level":"trace","@timestamp":"2022-03-02T23:53:24.060Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","reqId":"215419e0042e25919a4a32f6b74b1f4c","timeline":[[0.961036,"completePart intakeReq","socket hang up"]],"bytesWritten":0,"backoffDelayMs":928.5223783583905,"ecs":{"version":"1.6.0"},"error":{"type":"Error","message":"socket hang up","stack_trace":"Error: socket hang up\n    at connResetException (internal/errors.js:639:14)\n    at Socket.socketCloseListener (_http_client.js:449:25)\n    at Socket.emit (events.js:412:35)\n    at TCP.<anonymous> (net.js:686:12)\n    at TCP.callbackTrampoline (internal/async_hooks.js:130:17)"},"message":"conclude intake request: error"}
```

That last log line prettied up is:

```
[2022-03-02T23:53:24.060Z] TRACE (elastic-apm-node): conclude intake request: error
    event.module: "apmclient"
    reqId: "215419e0042e25919a4a32f6b74b1f4c"
    timeline: [[0.961036, "completePart intakeReq", "socket hang up"]]
    bytesWritten: 0
    backoffDelayMs: 928.5223783583905
    error: {
        "type": "Error",
        "message": "socket hang up",
        "stack_trace":
            Error: socket hang up
                at connResetException (internal/errors.js:639:14)
                at Socket.socketCloseListener (_http_client.js:449:25)
                at Socket.emit (events.js:412:35)
                at TCP.<anonymous> (net.js:686:12)
                at TCP.callbackTrampoline (internal/async_hooks.js:130:17)
    }
```

The client that the APM agent uses to talk to APM server (https://github.com/elastic/apm-nodejs-http-client) uses TCP keep-alive by default (via `keepAlive: true` to the Node.js core `http.Agent`). The `http.Agent` does not provide a mechanism to timeout kept-alive sockets, other than when it notices the connection is closed by the server -- i.e. dependent on the server's Keep-Alive timeout, whatever that is.

If a socket is re-used for a request near the time when the server is about to timeout the socket from its end, there is a race that can lead to the above "Error: socket hang up" being emitted as an "error" event on the request object.

### repro

I can reproduce this frequently in a Lambda environment (when the Lambda function is instrumented with the Node.js APM agent) when I space out invocations of the function by greater than 15s. The 15s is relevant because the current [Lambda extension](https://github.com/elastic/apm-aws-lambda) has a (default Go-lang) Keep-Alive timeout of 15s.

I can reproduce when spacing out invocations by *greater* than 15s (e.g. 20s, 35s, etc.). I believe this may be because while the Lambda VM is frozen neither the client (the APM agent) nor the server (the Lambda extension) is maintaining connections. When the VM is unfrozen there is perhaps a race between (a) the extension noticing that enough time has passed to drop its socket connection and (b) the APM agent sending a request to the extension on that re-used socket. But I'm guessing.

### links

There are a couple helpful links that describe others having his this scenario:

- https://connectreport.com/blog/tuning-http-keep-alive-in-node-js/
- See the "Resolving socket hang up and ECONNRESET errors in NodeJS" section in https://medium.com/ssense-tech/reduce-networking-errors-in-nodejs-23b4eb9f2d83

Other links:
- https://pkg.go.dev/net#Dialer shows that the default `KeepAlive time.Duration` is 15s.

### options

1. Turn off keep-alive usage in the Node.js APM agent. This would be unfortunate as it would impact performance (unmeasured) for higher-load cases.
2. Make sure that the APM agent's HTTP client times out kept-alive sockets sooner than the downstream server (APM server, or Lambda extension, or proxy) does. Because Node's core `http.Agent` doesn't offer this, we must do one of: (a) manually handle this, (b) use a different HTTP client or agent that handles this better. https://www.npmjs.com/package/agentkeepalive supports a `freeSocketTimeout` value that can be set. Undici (https://undici.nodejs.org/#/docs/api/Client) supports a `keepAliveTimeout` option for this. Both of these libraries default to 4s to handle the common case of being lower than the Node.js core `http.createServer` *server-side* `keepAliveTimeout` of 5s. As long as we can know a value that is less than the server-side keep alive timeout of any server that the Node.js APM agent is likely to talk to, then this is a good solution.
3. Handle retries when getting a ECONNRESET, as discussed here: https://nodejs.org/api/http.html#requestreusedsocket  While this is a good general solution, this would add significant complexity to the current streaming intake requests.

The fasted first stab is probably (2b), using agentkeepalive. Attempting to switch to Undici would be interesting because of its strong core Node backing and perf numbers.

**Comment by trentm** [comment]
Here is the patch to the http-client for option 1 (disabling keep-alive usage by the APM agent):

```
diff --git a/index.js b/index.js
index 1a30978..1f71fad 100644
--- a/index.js
+++ b/index.js
@@ -1299,7 +1308,8 @@ function getConfigRequestOptions (opts, agent) {

 function getBasicRequestOptions (method, defaultPath, headers, opts, agent) {
   return {
-    agent: agent,
+    // agent: agent,
+    agent: false, // XXX don't re-use a `http.Agent` for keep-alive
     rejectUnauthorized: opts.rejectUnauthorized !== false,
     ca: opts.serverCaCert,
     hostname: opts.serverUrl.hostname,
```

After some brief testing with this, I could no longer repro the "socket hang up" issue.  I.e. this is a possible fallback if the other options don't work.

**Comment by trentm** [comment]
Looking at option 2 (using agentkeepalive):

- They have a min node version of v8.0.0 (https://github.com/node-modules/agentkeepalive/blob/4.2.1/package.json#L54-L56). ❤️ 
- Deps are fairly small:
```
% npm ls 
...
└─┬ agentkeepalive@4.2.1
  ├─┬ debug@4.3.3
  │ └── ms@2.1.2
  ├── depd@1.1.2
  └─┬ humanize-ms@1.2.1
    └── ms@2.1.2 deduped

% du -sh node_modules
212K	node_modules
```
- It seems to *work* to me. See example below.

I'll get a PR together with this.


# example showing agentkeepalive usage

Here is an excerpt of a Lambda function VM log for two invocations of the function, with a 20s gap. This function was configured with `DEBUG=agentkeepalive` to show some internal processing of the agentkeepalive HTTP agent.

```
START RequestId: 49e8fa79-a6cc-43d3-b4f3-347b82228107 Version: $LATEST
...
{"log.level":"trace","@timestamp":"2022-03-07T20:59:40.450Z","log":{"logger":"elastic-apm-node"},"awsRequestId":"49e8fa79-a6cc-43d3-b4f3-347b82228107","ecs":{"version":"1.6.0"},"message":"lambda: fn start"}
...
2022-03-07T20:59:40.629Z agentkeepalive sock[0#localhost:8200:] create, timeout 30000ms
...
{"log.level":"trace","@timestamp":"2022-03-07T20:59:40.751Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","ecs":{"version":"1.6.0"},"message":"_signalLambdaEnd start"}
2022-03-07T20:59:40.752Z agentkeepalive sock[1#localhost:8200:] create, timeout 30000ms
2022-03-07T20:59:40.808Z agentkeepalive sock[0#localhost:8200:](requests: 1, finished: 1) free
...
{"log.level":"trace","@timestamp":"2022-03-07T20:59:41.088Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","durationMs":337.04886599991005,"ecs":{"version":"1.6.0"},"message":"signaled lambda invocation done"}
...
2022-03-07T20:59:41.108Z agentkeepalive sock[1#localhost:8200:](requests: 1, finished: 1) free
...
END RequestId: 49e8fa79-a6cc-43d3-b4f3-347b82228107
REPORT RequestId: 49e8fa79-a6cc-43d3-b4f3-347b82228107	Duration: 1180.40 ms	Billed Duration: 1181 ms	Memory Size: 128 MB	Max Memory Used: 100 MB	Init Duration: 749.97 ms
```

Here is the 20s gap that would fairly reliable result in the "socket hang up" error.
As soon as the second invocation starts, we see that the new HTTP agent times out the kept-alive sockets after the 4000ms timeout. This is what prevents the ECONNRESET / "socket hang up" errors.


```
START RequestId: 10f96eda-9e89-4286-abde-0b03e80ebefb Version: $LATEST
2022-03-07T21:00:01.271Z agentkeepalive sock[0#localhost:8200:](requests: 1, finished: 1) timeout after 4000ms, listeners 2, defaultTimeoutListenerCount 3, hasHttpRequest false, HttpRequest timeoutListenerCount 0
2022-03-07T21:00:01.271Z agentkeepalive timeout listeners: onTimeout, onTimeout
2022-03-07T21:00:01.271Z agentkeepalive sock[0#localhost:8200:] is free, destroy quietly
2022-03-07T21:00:01.271Z agentkeepalive sock[1#localhost:8200:](requests: 1, finished: 1) timeout after 4000ms, listeners 2, defaultTimeoutListenerCount 3, hasHttpRequest false, HttpRequest timeoutListenerCount 0
2022-03-07T21:00:01.271Z agentkeepalive timeout listeners: onTimeout, onTimeout
2022-03-07T21:00:01.271Z agentkeepalive sock[1#localhost:8200:] is free, destroy quietly
...
{"log.level":"trace","@timestamp":"2022-03-07T21:00:01.288Z","log":{"logger":"elastic-apm-node"},"awsRequestId":"10f96eda-9e89-4286-abde-0b03e80ebefb","ecs":{"version":"1.6.0"},"message":"lambda: fn start"}
...
```

(Aside, we also see that the intake request and the `?flushed=true` signal to the extension are using *different* sockets. That is perhaps something we could improve upon sometime later, but is unrelated to this issue.)

**Comment by trentm** [comment]
And with 1s spacing between function invocations we see the socket being re-used:

```
2022-03-07T21:36:23.516000+00:00 2022/03/07/[$LATEST]c6f2be16c1f44b40b02e2ef2e7901064 2022-03-07T21:36:23.516Z agentkeepalive sock[6#localhost:8200:] reset timeout to 30000ms
2022-03-07T21:36:23.516000+00:00 2022/03/07/[$LATEST]c6f2be16c1f44b40b02e2ef2e7901064 2022-03-07T21:36:23.516Z agentkeepalive sock[6#localhost:8200:](requests: 7, finished: 6) reuse on addRequest, timeout 30000ms
2022-03-07T21:36:23.516000+00:00 2022/03/07/[$LATEST]c6f2be16c1f44b40b02e2ef2e7901064 {"log.level":"trace","@timestamp":"2022-03-07T21:36:23.516Z","log":{"logger":"elastic-apm-node"},"event.module":"apmclient","reqId":"152027069048423fd44669fd1f63ee2f","ecs":{"version":"1.6.0"},"message":"intake request start"}
```

**Comment by trentm** [comment]
https://github.com/elastic/apm-nodejs-http-client/pull/179 added a fix to the http-client, which was published in elastic-apm-http-client@11.0.0.  Then #2586 updated the dep to elastic-apm-http-client@11.0.0, so this is now fixed and will be in the next agent release.
